File: voice_control/bridge/clients/csharp/tool_server.cs
// tool_server.cs
// Requires the NetMQ NuGet package.
// This is a server that exposes application functions to the Python client.

using System;
using System.Collections.Generic;
using System.Threading;
using NetMQ;
using NetMQ.Sockets;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

public class ToolServer : IDisposable
{
    private ResponseSocket _server;
    private readonly string _endpoint;
    private readonly string _authToken;
    private bool _isRunning = false;
    private Thread _serverThread;

    // --- RPC Method Dispatcher ---
    private readonly Dictionary<string, Func<JObject, JObject>> _rpcMethods;

    public ToolServer(string endpoint = "tcp://127.0.0.1:8080", string authToken = null)
    {
        _endpoint = endpoint;
        _authToken = authToken;
        _rpcMethods = new Dictionary<string, Func<JObject, JObject>>
        {
            {"move_player", MovePlayer},
            {"get_player_position", GetPlayerPosition},
            {"set_game_pause", SetGamePause},
            {"get_game_time", GetGameTime}
        };
    }

    #region Dummy Tool Implementations
    private JObject MovePlayer(JObject parameters)
    {
        var direction = parameters["direction"]?.ToString() ?? "unknown";
        Console.WriteLine($"[ToolServer] Moving player: {direction}");
        return new JObject {["status"] = "success", ["message"] = $"Moved {direction}"};
    }

    private JObject GetPlayerPosition(JObject parameters)
    {
        Console.WriteLine("[ToolServer] Getting player position");
        return new JObject {["x"] = 150, ["y"] = 250};
    }

    private JObject SetGamePause(JObject parameters)
    {
        var isPaused = parameters["is_paused"]?.ToObject<bool>() ?? false;
        Console.WriteLine($"[ToolServer] Setting game pause state to: {isPaused}");
        return new JObject {["status"] = "success"};
    }

    private JObject GetGameTime(JObject parameters)
    {
        Console.WriteLine("[ToolServer] Getting game time");
        return new JObject {["time"] = "3:00 PM"};
    }
    #endregion

    public void Start()
    {
        if (_isRunning) return;
        _isRunning = true;
        _serverThread = new Thread(ServerLoop);
        _serverThread.Start();
        Console.WriteLine($"[ToolServer] Listening on {_endpoint}");
        if (!string.IsNullOrEmpty(_authToken))
        {
            Console.WriteLine("[ToolServer] Authentication is enabled.");
        }
    }

    public void Stop()
    {
        _isRunning = false;
        // In a real app, you might need a more graceful shutdown
        if (_serverThread != null && _serverThread.IsAlive)
        {
            // NetMQ sockets can be tricky to close from another thread.
            // In a real Unity/Godot app, you'd integrate this into the main loop
            // or use async patterns instead of a blocking thread.
        }
        Console.WriteLine("[ToolServer] Stopped.");
    }

    private void ServerLoop()
    {
        using (var server = new ResponseSocket())
        {
            server.Bind(_endpoint);
            while (_isRunning)
            {
                if (server.TryReceiveFrameString(TimeSpan.FromSeconds(1), out var message))
                {
                    var response = _HandleRequest(message);
                    server.SendFrame(response);
                }
            }
        }
    }

    private string _HandleRequest(string requestJson)
    {
        JObject response;
        JObject request = null;
        try
        {
            request = JObject.Parse(requestJson);

            if (!string.IsNullOrEmpty(_authToken) && request["auth_token"]?.ToString() != _authToken)
            {
                throw new Exception("Authentication failed");
            }

            var methodName = request["method"]?.ToString();
            if (string.IsNullOrEmpty(methodName) || !_rpcMethods.ContainsKey(methodName))
            {
                throw new Exception("Method not found");
            }

            var result = _rpcMethods[methodName](request["params"] as JObject);
            response = new JObject
            {
                ["jsonrpc"] = "2.0",
                ["result"] = result,
                ["id"] = request["id"]
            };
        }
        catch (Exception ex)
        {
            response = new JObject
            {
                ["jsonrpc"] = "2.0",
                ["error"] = new JObject {["code"] = -32000, ["message"] = ex.Message},
                ["id"] = request?["id"]
            };
        }
        return response.ToString(Formatting.None);
    }

    public void Dispose()
    {
        Stop();
    }

    // Example Usage
    public static void Main(string[] args)
    {
        var authToken = Environment.GetEnvironmentVariable("TOOLS_AUTH_TOKEN");
        using (var server = new ToolServer(authToken: authToken))
        {
            server.Start();
            Console.WriteLine("ToolServer running. Press Enter to stop.");
            Console.ReadLine();
        }
    }
}

File: voice_control/bridge/clients/js/tool_server.js
// tool_server.js
// Requires the 'zeromq' npm package.
// This is a server that exposes application functions to the Python client.

const zmq = require("zeromq");

// --- Dummy Tool Implementations ---
function movePlayer(params) {
    console.log(`[ToolServer] Moving player: ${params.direction}`);
    return { status: "success", message: `Moved ${params.direction}` };
}

function getPlayerPosition(params) {
    console.log("[ToolServer] Getting player position");
    return { x: 50, y: 60 };
}

function setGamePause(params) {
    console.log(`[ToolServer] Setting game pause state to: ${params.is_paused}`);
    return { status: "success" };
}

function getGameTime(params) {
    console.log("[ToolServer] Getting game time");
    return { time: "10:00 AM" };
}

// --- RPC Method Dispatcher ---
const RPC_METHODS = {
    "move_player": movePlayer,
    "get_player_position": getPlayerPosition,
    "set_game_pause": setGamePause,
    "get_game_time": getGameTime,
};

class ToolServer {
    constructor(endpoint = "tcp://0.0.0.0:8080", authToken = null) {
        this.endpoint = endpoint;
        this.authToken = authToken;
        this.socket = new zmq.Reply();
        this.isRunning = false;
    }

    async start() {
        await this.socket.bind(this.endpoint);
        this.isRunning = true;
        console.log(`[ToolServer] Listening on ${this.endpoint}`);
        if (this.authToken) {
            console.log("[ToolServer] Authentication is enabled.");
        }

        while (this.isRunning) {
            try {
                const [msg] = await this.socket.receive();
                const request = JSON.parse(msg.toString());
                const response = this._handleRequest(request);
                await this.socket.send(JSON.stringify(response));
            } catch (e) {
                if (!this.isRunning) break;
                console.error("ZMQ receive error:", e);
            }
        }
    }

    stop() {
        this.isRunning = false;
        this.socket.close();
        console.log("[ToolServer] Stopped.");
    }

    _handleRequest(request) {
        try {
            if (this.authToken && request.auth_token !== this.authToken) {
                throw new Error("Authentication failed");
            }

            const method = RPC_METHODS[request.method];
            if (!method) {
                throw new Error("Method not found");
            }

            const result = method(request.params || {});
            return { jsonrpc: "2.0", result: result, id: request.id };

        } catch (e) {
            return {
                jsonrpc: "2.0",
                error: { code: -32000, message: e.message },
                id: request.id,
            };
        }
    }
}

// Example Usage
async function main() {
    const authToken = process.env.TOOLS_AUTH_TOKEN || null;
    const server = new ToolServer(undefined, authToken);

    console.log("ToolServer running. Press Ctrl+C to stop.");
    await server.start();
}

if (require.main === module) {
    main();
}

module.exports = ToolServer;